ACL2022/Models/are_models

```diff
@@ -0,0 +1,324 @@
+        self.max_seq_len =  max_seq_len
+        self.show_summary = show_summary
+        self.patience=patience
+        self.epochs = epochs
+        self.loss = loss
+        self.embedding_size = embedding_size
+        self.monitor_loss = monitor_loss
+        self.metrics = metrics
+        self.earlystop = tf.keras.callbacks.EarlyStopping(monitor=self.monitor_loss,
+                                                          patience=self.patience,
+                                                          verbose=1,
+                                                          restore_best_weights=True,
+                                                          mode="min"
+                                                         )
+        self.unk_token = "[unk]"
+        self.pad_token = "[pad]"
+
+  #dummy loss for att weights (ignores them)
+  def dummy_loss(self, y_true, y_pred):
+      return 0.0
+
+  def build(self, vocab_size):
+      inputs= Input(shape=(self.max_seq_len,), name="inputs")
+      x = Embedding(input_dim=vocab_size, output_dim=self.embedding_size, 
+                      input_length=self.max_seq_len, mask_zero=True, trainable=True)(inputs)
+      rnn = Bidirectional(LSTM(self.rnn_layer_size, return_sequences=True))(x)
+      for i in range(self.attention_hidden_layers): #-1):
+        if i == 0:
+          x = TimeDistributed(Dense(self.rnn_layer_size, activation = None))(rnn)
+        elif i == self.attention_hidden_layers-1: #last layer
+          x = TimeDistributed(Dense(1, activation = None))(x)
+        else:
+          x = TimeDistributed(Dense(self.rnn_layer_size, activation = None))(x) #hidden layer
+      att_weights = tensorflow.nn.softmax(x, axis=1)
+      weighted_represantation = tensorflow.math.multiply(att_weights, rnn, name=None)
+      weighted_represantation = tensorflow.reduce_sum(weighted_represantation, axis = 1)
+      pred = Dense(1, activation = None, name = 'classification')(weighted_represantation)
+      self.model = Model(inputs=inputs, outputs=[pred, tensorflow.squeeze(att_weights)])
+      self.model.compile(loss=[self.loss, self.dummy_loss] ,
+                    optimizer=tensorflow.keras.optimizers.Adam(learning_rate=self.lr),
+                    metrics=self.metrics)
+      if self.show_summary:
+          self.model.summary()
+
+  def create_vocab(self, tokenized_texts):
+    self.vocab = list(set([w for txt in tokenized_texts for w in txt]))
+    self.vocab_size = len(self.vocab) + 2
+    print("Vocab size: ", self.vocab_size)
+    self.w2i = {w: i+2 for i,w in enumerate(self.vocab)}
+    self.w2i[self.unk_token] = 1
+    self.w2i[self.pad_token] = 0
+    self.i2w = {i+2: self.w2i[w] for i,w in enumerate(self.vocab)}
+    self.i2w[1] = self.unk_token
+    self.i2w[0] = self.pad_token
+  
+  def to_sequences(self, tokenized_texts):
+        x = [[self.w2i[w] if w in self.w2i else self.w2i[self.unk_token] for w in t] for t in tokenized_texts]
```

Comment:
this is pretty incomprehensible, maybe a comment at least.

Reply:
+1 @alexandrosXe please check this: https://realpython.com/python-pep8/

Reply:
+1

Reply:
added comment